File: python/ransom_note.py
from collections import Counter


class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        note = Counter(ransomNote)
        magazine = Counter(magazine)
        return magazine >= note


# second solution
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        tally = [0 for _ in range(26)]
        for char in magazine:
            tally[ord(char) - 97] += 1

        for char in ransomNote:
            i = ord(char) - 97
            if tally[i] <= 0:
                return False
            tally[i] -= 1
        return True<Paste>

File: rust/src/ransom_note.rs
impl Solution {
    pub fn can_construct(ransom_note: String, magazine: String) -> bool {
        let mut map: Vec<u32> = Vec::with_capacity(26);
        for i in 0..26 {
            map.push(0);
        }
        // populate the "hash-map" with all the characters from magazine
        for c in magazine.chars() {
            // safe since we know all the characters will be 'a'..'z'
            let code_point: usize = ((c as u32) % 97) as usize;
            map[code_point] += 1;
        }
        // then iterate over ransom_note and subtract
        for c in ransom_note.chars() {
            let code_point: usize = ((c as u32) % 97) as usize;
            if map[code_point] <= 0 {
                return false;
            }
            map[code_point] -= 1;
        }
        // if any index would be negative, immediately return false
        return true;
    }
}
